MericCelik/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,266 @@
+    public ArrayList<BinaryTreeNode<T>> findAncestors(BinaryTreeNode<T> target){
+        ArrayList<BinaryTreeNode<T>> ancestors = new ArrayList<>();
+        BinaryTreeNode<T> temp = target;
+        while(temp.getParent()!=null){
+            ancestors.add(temp.getParent());
+            temp = temp.getParent();
+        }
+        return ancestors;
+    }
+
+    /**Calls find() and findAncestors() method and print the items of returned arraylist.
+     * Exceptions are handled in find() method and it returns the node with given element.
+     * If the target node is the root, it prints a warning message to the stardard output.
+     * Complexities: depends on find() and findsAncestors() methods.
+     * @param element : element of node whose ancestors are found.
+     * @return nothing.
+     * */
+    public void printAncestors(T element){
+
+        BinaryTreeNode<T> targetnode = this.find(element);
+
+        if(targetnode.equals(this.root)){
+            System.out.println("Target node is the root of the tree, it has no parent!");
+            return;
+        }
+
+        ArrayList<BinaryTreeNode<T>> ancestors = findAncestors(targetnode);
+        int ancestor_count = ancestors.size();
+
+        for(int i=0; i<ancestor_count;i++)
+            System.out.print(ancestors.get(i).getKey() + " ");
+        System.out.println();
+    }
+
+    /**Find the lowest common ancestor of the nodes with given parameters. Firstly, the method calls find() method with first
+     * parameter to find the corresponding node. After that, it choose this node as "root", and calls findNodeWithDFS()
+     * method to find if the other parameter(elem2) is found in the subtree with this "root". If not, findNodeWithDFS() function
+     * returns null and the search goes on. The new "root" is the "root"s parent and the same process is done. At some point,
+     * if the function finds "elem2", it returns current "root" as lowest common ancestor.
+     * The order of the nodes (or parameters) is not important, so we easily picked the first parameter's node as a root.
+     * Complexities: Average O(N^2) since we are doing dfs for every parent of the "root",
+     *               Best O(1) if if two elements are the same.
+     *               Worst O(N)
+     * @param elem1 : element of the first node
+     * @param elem2 : element of the other node
+     * @throws NullPointerException if one or both of the elements are null.
+     * @return lowest common ancestor node, null otherwise
+     * */
+    public BinaryTreeNode<T> lowestCommonAncestor(T elem1, T elem2){
+
+        if(elem1==null)
+            throw new NullPointerException("First argument is invalid!");
+        if(elem2==null)
+            throw new NullPointerException("Second argument is invalid!");
+
+        BinaryTreeNode<T> node1 = this.find(elem1);
+        BinaryTreeNode<T> found = null;
+
+        while(found==null && node1 !=null){
+            found = this.findNodeWithDFS(node1, elem2);
```

Comment:
The way to do this in O(N) here is to check if we've already visited that sub-tree :) just one little change and such a big difference!
So what we'd do here with this loop is something like this (there could be bugs, I'm writing it from the top of my head right now :stuck_out_tongue:) :

```
BinaryTreeNode<T> node1 = this.find(elem1);
BinaryTreeNode<T> visitedChild = null; // remember where we came from
BinaryTreeNode<T> found = null;

while (found == null && node1 != null) {
    // check the left sub-tree if we haven't just came from there.
    if (visitedChild == null || visitedChild != node1.getLeftChild()) {
        found = this.findNodeWithDFS(node1.getLeftChild(), elem2);
    }
    // if we haven't found the node and we haven't come from the right sub-tree, look there
    if (found == null && (visitedChild == null || visitedChild != node1.getRightChild()) {
        found = this.findNodeWithDFS(node.getRightChild(), elem2);
    }
    if (found != null) {
        break;
    }
    visitedChild = node1;
    node1 = node1.getParent();
}
```

Reply:
Amazing idea, thank you! I implemented but it could not find lca when the first node is in the second node's subtree. So I added this condition too and it works pretty well. I hope you like it! :relaxed: